### PR TITLE
feat: make subtitleeditparser generic

### PR DESCRIPTION
**Pull request by cvium**
**Changes**
Instead of hard coding and wrapping every `SubtitleFormat` sub class, this introduces a generic parser that tries all available `SubtitleFormat` based on extension.

**Issues**
Supercedes #8086

### CODE REVIEW

**Review by crobibero — APPROVED** [review]

File: MediaBrowser.MediaEncoding/Subtitles/SubtitleEditParser.cs [hunk 1]
```diff
@@ -57,5 +85,94 @@ public SubtitleTrackInfo Parse(Stream stream, CancellationToken cancellationToke
...
+            var assembly = Assembly.GetAssembly(typeof(SubtitleFormat));
+            if (assembly == null)
+            {
+                _logger.LogError("Missing assembly containing {SubtitleFormatName}", nameof(SubtitleFormat));
+                return GetFallbackSubtitleFormats();
+            }
+
+            foreach (var type in assembly.GetTypes())
+            {
+                if (!type.IsSubclassOf(typeof(SubtitleFormat)))
```
Comment (Bond-009) [inline MediaBrowser.MediaEncoding/Subtitles/SubtitleEditParser.cs:105]: Should we check to make sure it isn't abstract?
```suggestion
                if (!type.IsSubclassOf(typeof(SubtitleFormat)) || type.IsAbstract)
```

File: MediaBrowser.MediaEncoding/Subtitles/SubtitleEditParser.cs [hunk 2]
```diff
@@ -57,5 +85,94 @@ public SubtitleTrackInfo Parse(Stream stream, CancellationToken cancellationToke
...
         }
+
+        /// <inheritdoc />
+        public bool SupportsFileExtension(string fileExtension)
+            => _subtitleFormats.ContainsKey(fileExtension);
+
+        private IEnumerable<SubtitleFormat> GetSubtitleFormats()
+        {
+            var subtitleFormats = new List<SubtitleFormat>();
+            var assembly = Assembly.GetAssembly(typeof(SubtitleFormat));
```
Comment (Bond-009) [inline MediaBrowser.MediaEncoding/Subtitles/SubtitleEditParser.cs:96]:
```suggestion
            var assembly = typeof(SubtitleFormat).Assembly;
```
https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getassembly?view=net-7.0#remarks

File: MediaBrowser.MediaEncoding/Subtitles/SubtitleEditParser.cs [hunk 3]
```diff
@@ -57,5 +85,94 @@ public SubtitleTrackInfo Parse(Stream stream, CancellationToken cancellationToke
...
+            if (assembly == null)
```
Comment (Bond-009) [inline MediaBrowser.MediaEncoding/Subtitles/SubtitleEditParser.cs:97]: Is this case ever hit?

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: MediaBrowser.MediaEncoding/Subtitles/SubtitleEditParser.cs (hunk 3 above)
Comment (cvium) [reply MediaBrowser.MediaEncoding/Subtitles/SubtitleEditParser.cs:97]: I doubt it, but it's nullable 🤷🏼

Comment (Bond-009) [reply MediaBrowser.MediaEncoding/Subtitles/SubtitleEditParser.cs:97]: Not anymore ;)

Comment (cvium) [reply MediaBrowser.MediaEncoding/Subtitles/SubtitleEditParser.cs:97]: You can remove it then

File: MediaBrowser.MediaEncoding/Subtitles/SubtitleEditParser.cs [hunk 4]
```diff
@@ -57,5 +85,89 @@ public SubtitleTrackInfo Parse(Stream stream, CancellationToken cancellationToke
...
+                new Csv3(),
+                new Csv4(),
+                new Csv5(),
+                new Ebu(),
+                new NetflixImsc11Japanese(),
+                new NetflixTimedText(),
+                new QuickTimeText(),
+                new RealTime(),
+                new SmpteTt2052()
+            };
```
Comment (Bond-009) [inline MediaBrowser.MediaEncoding/Subtitles/SubtitleEditParser.cs:171]:
```suggestion
```
Isn't this dead code now?

**Review by Bond-009 — CHANGES_REQUESTED** [review]

**Review by Bond-009 — APPROVED** [review]

**Pull request merged**